File: app/src/commonTest/java/org/mifos/mobilebanking/FakeRemoteDataSource.java
package org.mifos.mobilebanking;

import com.google.gson.reflect.TypeToken;

import org.mifos.mobilebanking.models.Charge;
import org.mifos.mobilebanking.models.Page;
import org.mifos.mobilebanking.models.Transaction;
import org.mifos.mobilebanking.models.User;
import org.mifos.mobilebanking.models.accounts.loan.LoanAccount;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobilebanking.models.accounts.savings.SavingsWithAssociations;
import org.mifos.mobilebanking.models.beneficiary.Beneficiary;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryPayload;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryUpdatePayload;
import org.mifos.mobilebanking.models.client.Client;
import org.mifos.mobilebanking.models.client.ClientAccounts;
import org.mifos.mobilebanking.models.guarantor.GuarantorPayload;
import org.mifos.mobilebanking.models.guarantor.GuarantorTemplatePayload;
import org.mifos.mobilebanking.models.payload.LoansPayload;
import org.mifos.mobilebanking.models.payload.TransferPayload;
import org.mifos.mobilebanking.models.register.RegisterPayload;
import org.mifos.mobilebanking.models.register.UserVerify;
import org.mifos.mobilebanking.models.templates.account.AccountOptionsTemplate;
import org.mifos.mobilebanking.models.templates.beneficiary.BeneficiaryTemplate;
import org.mifos.mobilebanking.models.templates.loans.LoanTemplate;
import org.mifos.mobilebanking.models.templates.savings.SavingsAccountTemplate;

import java.util.List;

/**
 * Created by dilpreet on 26/6/17.
 */

public class FakeRemoteDataSource {

    private static TestDataFactory mTestDataFactory = new TestDataFactory();

    public static ClientAccounts getClientAccounts() {
        return mTestDataFactory.getListTypePojo(new TypeToken<ClientAccounts>() {
        }, FakeJsonName.CLIENT_ACCOUNTS_JSON);
    }

    public static ClientAccounts getClientSavingsAccount() {
        return mTestDataFactory.getObjectTypePojo(ClientAccounts.class,
                FakeJsonName.CLIENT_SAVINGS_ACCOUNT_JSON);
    }

    public static ClientAccounts getClientLoanAccount() {
        return mTestDataFactory.getObjectTypePojo(ClientAccounts.class,
                FakeJsonName.CLIENT_LOAN_ACCOUNT_JSON);
    }

    public static ClientAccounts getClientShareAccount() {
        return mTestDataFactory.getObjectTypePojo(ClientAccounts.class,
                FakeJsonName.CLIENT_SHARE_ACCOUNT_JSON);
    }

    public static List<Beneficiary> getBeneficiaries() {
        return mTestDataFactory.getListTypePojo(new TypeToken<List<Beneficiary>>() {
        }, FakeJsonName.BENEFICIARIES_JSON);
    }

    public static BeneficiaryTemplate getBeneficiaryTemplate() {
        return mTestDataFactory.getObjectTypePojo(BeneficiaryTemplate.class,
                FakeJsonName.BENEFICIARY_TEMPLATE_JSON);
    }

    public static BeneficiaryPayload beneficiaryPayload() {
        return mTestDataFactory.getObjectTypePojo(BeneficiaryPayload.class,
                FakeJsonName.BENEFICIARY_PAYLOAD_JSON);
    }

    public static BeneficiaryUpdatePayload beneficiaryUpdatePayload() {
        return mTestDataFactory.getObjectTypePojo(BeneficiaryUpdatePayload.class,
                FakeJsonName.BENEFICIARY_UPDATE_PAYLOAD_JSON);
    }

    public static LoanAccount getLoanAccount() {
        return mTestDataFactory.getObjectTypePojo(LoanAccount.class,
                FakeJsonName.LOAN_ACCOUNT_JSON);
    }

    public static LoanTemplate getLoanTemplate() {
        return mTestDataFactory.getObjectTypePojo(LoanTemplate.class,
                FakeJsonName.LOAN_TEMPLATE_JSON);
    }

    public static LoansPayload getLoansPayload() {
        return mTestDataFactory.getObjectTypePojo(LoansPayload.class,
                FakeJsonName.LOAN_PAYLOAD_JSON);
    }

    public static LoanTemplate getLoanTemplateByTemplate() {
        return mTestDataFactory.getObjectTypePojo(LoanTemplate.class,
                FakeJsonName.LOAN_TEMPLATE_BY_PRODUCT_JSON);
    }

    public static LoanWithAssociations getLoanAccountWithTransaction() {
        return mTestDataFactory.getObjectTypePojo(LoanWithAssociations.class,
                FakeJsonName.LOAN_ACCOUNT_WITH_TRANSACTIONS_JSON);
    }

    public static LoanWithAssociations getLoanAccountWithEmptyTransaction() {
        return mTestDataFactory.getObjectTypePojo(LoanWithAssociations.class,
                FakeJsonName.LOAN_ACCOUNT_WITH_EMPTY_TRANSACTIONS_JSON);
    }

    public static LoanWithAssociations getLoanAccountRepaymentSchedule() {
        return mTestDataFactory.getObjectTypePojo(LoanWithAssociations.class,
                FakeJsonName.LOAN_ACCOUNT_WITH_REPAYMENT_SCHEDULE_JSON);
    }

    public static LoanWithAssociations getLoanAccountEmptyRepaymentSchedule() {
        return mTestDataFactory.getObjectTypePojo(LoanWithAssociations.class,
                FakeJsonName.LOAN_ACCOUNT_WITH_EMPTY_REPAYMENT_SCHEDULE_JSON);
    }

    public static User getUser() {
        return mTestDataFactory.getObjectTypePojo(User.class,
                FakeJsonName.USER_JSON);
    }

    public static Page<Client> getClients() {
        return mTestDataFactory.getListTypePojo(new TypeToken<Page<Client>>() {
        }, FakeJsonName.CLIENTS_JSON);
    }

    public static Page<Client> getNoClients() {
        return mTestDataFactory.getListTypePojo(new TypeToken<Page<Client>>() {
        }, FakeJsonName.CLIENTS_NOT_FOUND_JSON);
    }

    public static Page<Transaction> getTransactions() {
        return mTestDataFactory.getListTypePojo(new TypeToken<Page<Transaction>>() {
        }, FakeJsonName.TRANSACTIONS_JSON);
    }

    public static SavingsWithAssociations getSavingsWithAssociations() {
        return mTestDataFactory.getObjectTypePojo(SavingsWithAssociations.class,
                FakeJsonName.SAVINGS_WITH_ASSOCIATIONS);
    }

    public static AccountOptionsTemplate getAccountOptionsTemplate() {
        return mTestDataFactory.getObjectTypePojo(AccountOptionsTemplate.class,
                FakeJsonName.ACCOUNT_OPTION_TEMPLATE);
    }

    public static TransferPayload getTransferPayload() {
        return mTestDataFactory.getObjectTypePojo(TransferPayload.class,
                FakeJsonName.TRANFER_PAYLOAD_JSON);
    }


    public static RegisterPayload getRegisterPayload() {
        return mTestDataFactory.getObjectTypePojo(RegisterPayload.class,
                FakeJsonName.REGISTER);
    }

    public static Page<Charge> getCharge() {
        return mTestDataFactory.getListTypePojo(new TypeToken<Page<Charge>>() {
        }, FakeJsonName.CHARGE);
    }

    public static List<Charge> getSavingsCharge() {
        return mTestDataFactory.getListTypePojo(new TypeToken<List<Charge>>() {
        }, FakeJsonName.SAVING_CHARGE);
    }

    public static List<Charge> getLoanCharge() {
        return mTestDataFactory.getListTypePojo(new TypeToken<List<Charge>>() {
        }, FakeJsonName.LOAN_CHARGE);
    }

    public static UserVerify getUserVerify() {
        return mTestDataFactory.getObjectTypePojo(UserVerify.class,
                FakeJsonName.USER_VERIFY_JSON);
    }

    public static GuarantorTemplatePayload getGuarantorTemplatePayload() {
        return mTestDataFactory.getObjectTypePojo(GuarantorTemplatePayload.class,
                FakeJsonName.GUARANTOR_TEMPLATE);
    }

    public static List<GuarantorPayload> getGuarantorsList() {
        return mTestDataFactory.getListTypePojo(new TypeToken<List<GuarantorPayload>>() {
        }, FakeJsonName.GUARANTOR_LIST);
    }

    public static SavingsAccountTemplate getSavingAccountApplicationTemplate() {
        return mTestDataFactory.getObjectTypePojo(SavingsAccountTemplate.class,
                FakeJsonName.SAVINGS_ACCOUNT_TEMPLATE);

    }
}

File: core/testing/src/main/java/org/mifos/mobile/core/testing/util/FakeRemoteDataSource.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-mobile/blob/master/LICENSE.md
 */
package org.mifos.mobile.core.testing.util

import com.google.gson.reflect.TypeToken
import org.mifos.mobile.core.database.entity.ChargeEntity
import org.mifos.mobile.core.model.entity.Page
import org.mifos.mobile.core.model.entity.Transaction
import org.mifos.mobile.core.model.entity.UpdatePasswordPayload
import org.mifos.mobile.core.model.entity.User
import org.mifos.mobile.core.model.entity.accounts.loan.LoanAccount
import org.mifos.mobile.core.model.entity.accounts.loan.LoanWithAssociations
import org.mifos.mobile.core.model.entity.accounts.savings.SavingsWithAssociations
import org.mifos.mobile.core.model.entity.beneficiary.Beneficiary
import org.mifos.mobile.core.model.entity.beneficiary.BeneficiaryPayload
import org.mifos.mobile.core.model.entity.beneficiary.BeneficiaryUpdatePayload
import org.mifos.mobile.core.model.entity.client.Client
import org.mifos.mobile.core.model.entity.client.ClientAccounts
import org.mifos.mobile.core.model.entity.guarantor.GuarantorPayload
import org.mifos.mobile.core.model.entity.guarantor.GuarantorTemplatePayload
import org.mifos.mobile.core.model.entity.payload.LoansPayload
import org.mifos.mobile.core.model.entity.payload.LoginPayload
import org.mifos.mobile.core.model.entity.payload.TransferPayload
import org.mifos.mobile.core.model.entity.register.RegisterPayload
import org.mifos.mobile.core.model.entity.register.UserVerify
import org.mifos.mobile.core.model.entity.templates.account.AccountOptionsTemplate
import org.mifos.mobile.core.model.entity.templates.beneficiary.BeneficiaryTemplate
import org.mifos.mobile.core.model.entity.templates.loans.LoanTemplate
import org.mifos.mobile.core.model.entity.templates.savings.SavingsAccountTemplate

object FakeRemoteDataSource {
    private val mTestDataFactory = TestDataFactory()
    val clientAccounts: ClientAccounts
        get() = mTestDataFactory.getListTypePojo(
            object :
                TypeToken<ClientAccounts?>() {},
            FakeJsonName.CLIENT_ACCOUNTS_JSON,
        )
    val currentClient: Client
        get() = mTestDataFactory.getObjectTypePojo(
            Client::class.java,
            FakeJsonName.CLIENT_JSON,
        )
    val clientSavingsAccount: ClientAccounts
        get() = mTestDataFactory.getObjectTypePojo(
            ClientAccounts::class.java,
            FakeJsonName.CLIENT_SAVINGS_ACCOUNT_JSON,
        )
    val clientLoanAccount: ClientAccounts
        get() = mTestDataFactory.getObjectTypePojo(
            ClientAccounts::class.java,
            FakeJsonName.CLIENT_LOAN_ACCOUNT_JSON,
        )
    val clientShareAccount: ClientAccounts
        get() = mTestDataFactory.getObjectTypePojo(
            ClientAccounts::class.java,
            FakeJsonName.CLIENT_SHARE_ACCOUNT_JSON,
        )
    val beneficiaries: List<Beneficiary>
        get() = mTestDataFactory.getListTypePojo(
            object :
                TypeToken<List<Beneficiary>?>() {},
            FakeJsonName.BENEFICIARIES_JSON,
        )
    val beneficiaryTemplate: BeneficiaryTemplate
        get() = mTestDataFactory.getObjectTypePojo(
            BeneficiaryTemplate::class.java,
            FakeJsonName.BENEFICIARY_TEMPLATE_JSON,
        )

    fun beneficiaryPayload(): BeneficiaryPayload {
        return mTestDataFactory.getObjectTypePojo(
            BeneficiaryPayload::class.java,
            FakeJsonName.BENEFICIARY_PAYLOAD_JSON,
        )
    }

    fun beneficiaryUpdatePayload(): BeneficiaryUpdatePayload {
        return mTestDataFactory.getObjectTypePojo(
            BeneficiaryUpdatePayload::class.java,
            FakeJsonName.BENEFICIARY_UPDATE_PAYLOAD_JSON,
        )
    }

    val loanAccount: LoanAccount
        get() = mTestDataFactory.getObjectTypePojo(
            LoanAccount::class.java,
            FakeJsonName.LOAN_ACCOUNT_JSON,
        )
    val loanTemplate: LoanTemplate
        get() = mTestDataFactory.getObjectTypePojo(
            LoanTemplate::class.java,
            FakeJsonName.LOAN_TEMPLATE_JSON,
        )
    val loansPayload: LoansPayload
        get() = mTestDataFactory.getObjectTypePojo(
            LoansPayload::class.java,
            FakeJsonName.LOAN_PAYLOAD_JSON,
        )
    val loanTemplateByTemplate: LoanTemplate
        get() = mTestDataFactory.getObjectTypePojo(
            LoanTemplate::class.java,
            FakeJsonName.LOAN_TEMPLATE_BY_PRODUCT_JSON,
        )
    val loanAccountWithTransaction: LoanWithAssociations
        get() = mTestDataFactory.getObjectTypePojo(
            LoanWithAssociations::class.java,
            FakeJsonName.LOAN_ACCOUNT_WITH_TRANSACTIONS_JSON,
        )
    val loanAccountWithEmptyTransaction: LoanWithAssociations
        get() = mTestDataFactory.getObjectTypePojo(
            LoanWithAssociations::class.java,
            FakeJsonName.LOAN_ACCOUNT_WITH_EMPTY_TRANSACTIONS_JSON,
        )
    val loanAccountRepaymentSchedule: LoanWithAssociations
        get() = mTestDataFactory.getObjectTypePojo(
            LoanWithAssociations::class.java,
            FakeJsonName.LOAN_ACCOUNT_WITH_REPAYMENT_SCHEDULE_JSON,
        )
    val loanAccountEmptyRepaymentSchedule: LoanWithAssociations
        get() = mTestDataFactory.getObjectTypePojo(
            LoanWithAssociations::class.java,
            FakeJsonName.LOAN_ACCOUNT_WITH_EMPTY_REPAYMENT_SCHEDULE_JSON,
        )
    val user: User
        get() = mTestDataFactory.getObjectTypePojo(
            User::class.java,
            FakeJsonName.USER_JSON,
        )
    val clients: Page<Client?>?
        get() = mTestDataFactory.getListTypePojo<Page<Client?>?>(
            object :
                TypeToken<Page<Client?>?>() {},
            FakeJsonName.CLIENTS_JSON,
        )
    val noClients: Page<Client?>?
        get() = mTestDataFactory.getListTypePojo<Page<Client?>?>(
            object :
                TypeToken<Page<Client?>?>() {},
            FakeJsonName.CLIENTS_NOT_FOUND_JSON,
        )
    val transactions: Page<Transaction?>?
        get() = mTestDataFactory.getListTypePojo<Page<Transaction?>?>(
            object :
                TypeToken<Page<Transaction?>?>() {},
            FakeJsonName.TRANSACTIONS_JSON,
        )
    val savingsWithAssociations: SavingsWithAssociations
        get() = mTestDataFactory.getObjectTypePojo(
            SavingsWithAssociations::class.java,
            FakeJsonName.SAVINGS_WITH_ASSOCIATIONS,
        )
    val accountOptionsTemplate: AccountOptionsTemplate
        get() = mTestDataFactory.getObjectTypePojo(
            AccountOptionsTemplate::class.java,
            FakeJsonName.ACCOUNT_OPTION_TEMPLATE,
        )
    val transferPayload: TransferPayload
        get() = mTestDataFactory.getObjectTypePojo(
            TransferPayload::class.java,
            FakeJsonName.TRANFER_PAYLOAD_JSON,
        )
    val registerPayload: RegisterPayload
        get() = mTestDataFactory.getObjectTypePojo(
            RegisterPayload::class.java,
            FakeJsonName.REGISTER,
        )
    val loginPayload: LoginPayload
        get() = mTestDataFactory.getObjectTypePojo(
            LoginPayload::class.java,
            FakeJsonName.LOGIN,
        )
    val charge: Page<ChargeEntity?>?
        get() = mTestDataFactory.getListTypePojo<Page<ChargeEntity?>?>(
            object :
                TypeToken<Page<ChargeEntity?>?>() {},
            FakeJsonName.CHARGE,
        )
    val savingsCharge: List<ChargeEntity>
        get() = mTestDataFactory.getListTypePojo(
            object :
                TypeToken<List<ChargeEntity>?>() {},
            FakeJsonName.SAVING_CHARGE,
        )
    val loanCharge: List<ChargeEntity>
        get() = mTestDataFactory.getListTypePojo(
            object :
                TypeToken<List<ChargeEntity>?>() {},
            FakeJsonName.LOAN_CHARGE,
        )
    val userVerify: UserVerify
        get() = mTestDataFactory.getObjectTypePojo(
            UserVerify::class.java,
            FakeJsonName.USER_VERIFY_JSON,
        )
    val guarantorTemplatePayload: GuarantorTemplatePayload
        get() = mTestDataFactory.getObjectTypePojo(
            GuarantorTemplatePayload::class.java,
            FakeJsonName.GUARANTOR_TEMPLATE,
        )
    val guarantorsList: List<GuarantorPayload>
        get() = mTestDataFactory.getListTypePojo(
            object :
                TypeToken<List<GuarantorPayload>?>() {},
            FakeJsonName.GUARANTOR_LIST,
        )
    val savingAccountApplicationTemplate: SavingsAccountTemplate
        get() = mTestDataFactory.getObjectTypePojo(
            SavingsAccountTemplate::class.java,
            FakeJsonName.SAVINGS_ACCOUNT_TEMPLATE,
        )
    val updatePasswordPayload: UpdatePasswordPayload
        get() = mTestDataFactory.getObjectTypePojo(
            UpdatePasswordPayload::class.java,
            FakeJsonName.UPDATE_PASSWORD_PAYLOAD_JSON,
        )
}
